### PR TITLE
repo: leverage IdIndex for change id resolution and divergent template

### PR DESCRIPTION
**Pull request by yuja**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [x] I have added tests to cover my changes

### CODE REVIEW

**Review by martinvonz — APPROVED** [review]
Looks good. @ilyagr, do you also want to take a look?

File: lib/src/repo.rs [hunk 1]
```diff
@@ -1252,7 +1302,59 @@ mod tests {
...
+            id_index.resolve_prefix_with(&HexPrefix::new("000").unwrap(), |&v| v),
+            PrefixResolution::SingleMatch(vec![0]),
+        );
+        assert_eq!(
+            id_index.resolve_prefix_with(&HexPrefix::new("0001").unwrap(), |&v| v),
+            PrefixResolution::NoMatch,
+        );
+        assert_eq!(
+            sorted(id_index.resolve_prefix_with(&HexPrefix::new("009").unwrap(), |&v| v)),
+            PrefixResolution::SingleMatch(vec![1, 2]),
```
Comment (yuja) [inline lib/src/repo.rs:1385]: Added explicit `sort()` here because I use `.sort_unstable_by()` to sort entries by key, though there wouldn't be no runtime randomness.

File: lib/src/repo.rs [hunk 2]
```diff
@@ -1220,6 +1234,42 @@ where
...
+    pub fn resolve_prefix_range<'a: 'b, 'b>(
+        &'a self,
+        prefix: &'b HexPrefix,
+    ) -> impl Iterator<Item = (&'a K, &'a V)> + 'b {
+        let (bytes_prefix, min_bytes_prefix) = prefix.bytes_prefixes::<K>();
+        let pos = self.0.partition_point(|(k, _)| k < &min_bytes_prefix);
+        self.0[pos..]
+            .iter()
+            .take_while(move |(k, _)| k.as_bytes().starts_with(bytes_prefix.as_bytes()))
+            .filter(|(k, _)| prefix.matches(k))
```
Comment (ilyagr) [inline lib/src/repo.rs:1314]: (See above) Why not just `take_while(|k| prefix.matches(k))`?

File: lib/src/repo.rs [hunk 3]
```diff
@@ -1220,6 +1234,42 @@ where
...
+        }
+    }
+
+    /// Iterates over entries with the given prefix.
+    pub fn resolve_prefix_range<'a: 'b, 'b>(
+        &'a self,
+        prefix: &'b HexPrefix,
+    ) -> impl Iterator<Item = (&'a K, &'a V)> + 'b {
+        let (bytes_prefix, min_bytes_prefix) = prefix.bytes_prefixes::<K>();
+        let pos = self.0.partition_point(|(k, _)| k < &min_bytes_prefix);
```
Comment (ilyagr) [inline lib/src/repo.rs:1310]: (See above) Would it be possible to encapsulate the distinciton between `bytes_prefix` and `min_bytes_prefix` inside HexPrefix, so that this comparison is just with a HexPrefix object?

File: lib/src/repo.rs [hunk 4]
```diff
@@ -1220,6 +1234,42 @@ where
...
+            PrefixResolution::NoMatch
+        }
+    }
+
+    /// Iterates over entries with the given prefix.
+    pub fn resolve_prefix_range<'a: 'b, 'b>(
+        &'a self,
+        prefix: &'b HexPrefix,
+    ) -> impl Iterator<Item = (&'a K, &'a V)> + 'b {
+        let (bytes_prefix, min_bytes_prefix) = prefix.bytes_prefixes::<K>();
```
Comment (ilyagr) [inline lib/src/repo.rs:1309]: Optional: I find this a bit hard to follow. It would help if it was documented what properties `bytes_prefix` and `min_bytes_prefix` satisfy. It would help even more if all of this optimization was hidden inside the implementation of HexPrefix.

The two suggestions below are the first things that came to mind, though I'm not sure if either is actually doable / a good idea.

It's fine to leave this as is, an think about whether this can be made clearer later.

File: lib/src/repo.rs [hunk 5]
```diff
@@ -660,6 +670,31 @@ impl MutableRepo {
         (self.index, self.view.into_inner())
     }
 
+    pub fn resolve_change_id_prefix<'a>(
```
Comment (ilyagr) [inline lib/src/repo.rs:683]: Optional: It might be nice to leave a comment explaining that this is different from and much slower than the ReadonlyRepo implementation

**Review by ilyagr — APPROVED** [review]
This looks good to me. Nice!

I left some about wording or various places where I wonder if things could be made clearer. In many cases, I haven't made up mind about that myself, so all of them are optional.

File: lib/src/repo.rs (hunk 2 above)
Comment (ilyagr) [reply lib/src/repo.rs:1314]: For some reason, the comment this is related to (https://github.com/martinvonz/jj/pull/1124#discussion_r1087424756) ended up below this one.

File: lib/src/repo.rs (hunk 3 above)
Comment (ilyagr) [reply lib/src/repo.rs:1310]: For some reason, the comment this is related to (https://github.com/martinvonz/jj/pull/1124#discussion_r1087424756) ended up below this one.

File: lib/src/repo.rs [hunk 6]
```diff
@@ -90,6 +92,26 @@ impl<'a> RepoRef<'a> {
         }
     }
 
+    pub fn resolve_change_id(&self, change_id: &ChangeId) -> Option<Vec<IndexEntry<'a>>> {
+        // Replace this if we added more efficient lookup method.
+        let prefix = HexPrefix::from_bytes(change_id.as_bytes());
+        match self.resolve_change_id_prefix(&prefix) {
+            PrefixResolution::NoMatch => None,
+            PrefixResolution::SingleMatch(entries) => Some(entries),
+            PrefixResolution::AmbiguousMatch => panic!("change_id should be unambiguous"),
```
[the remark this answers is not in the document]
Comment (yuja) [reply lib/src/repo.rs]: Sounds good. Added "complete".

File: lib/src/repo.rs (hunk 2 above)
Comment (yuja) [reply lib/src/repo.rs:1314]: Good catch. I copied the logic from index.rs, but `bytes_prefix` test wouldn't be needed.
Maybe it's there because `prefix.matches()` was slower before?

I'll make another PR for this.

File: lib/src/repo.rs (hunk 4 above)
Comment (yuja) [reply lib/src/repo.rs:1309]: I'll add doc comment to `HexPrefix` in another PR.

File: lib/src/repo.rs (hunk 5 above)
Comment (yuja) [reply lib/src/repo.rs:683]: I think the TODO comment copied from revset.rs implies that. It's not perfect, but we can easily deduce that.
